File: 0152-maximum-product-subarray/0152-maximum-product-subarray.js
/**
 * @param {number[]} nums
 * @return {number}
 */
var maxProduct = function(nums) {

    let res = Math.max(...nums)
    let curMin = 1
    let curMax = 1
    
    for (let num of nums ){
        
        if (num === 0){
            curMin = 1
            curMax = 1
            continue 
        }
        
        let temp = num * curMin 
        curMin = Math.min(temp, num * curMax, num) 
        curMax = Math.max (num * curMax, temp, num )
        
        res = Math.max(res, curMax)
    }
    
    return res 
};

File: solution_mixed_level/0152-maximum-product-subarray/0152-maximum-product-subarray.py
class Solution:
    def maxProduct(self, nums: List[int]) -> int:
        res = max(nums) # nums = [-1, -3, 0, 60] 같은 경우를 위해서 
        curMin, curMax = 1, 1 
        
        for n in nums: 
            if n == 0:
                curMin, curMax = 1, 1
                continue 
            temp = curMin * n 
            curMin = min(n * curMax, n* curMin, n)
            curMax = max(n * curMax, temp, n)
          
            res = max(res, curMax)
        
        return res 